microscopium/_util.py

```diff
@@ -1,6 +1,82 @@
 import numpy as np
 import numbers
 
+
+def generate_spiral(dim, direction, clockwise=False):
+    """Generate an dim * dim of sequental integers in a spiral pattern.
+
+    Many HCS systems use a spiral pattern to capture multiple fields in
+    for each well. This function generates a spiral pattern of integers
+    to be used by the montaging functions.
+
+    Parameters
+    ----------
+    dim : int
+        The dimension of the spiral. The function will return a dim * dim
+        array.
+    direction : str
+        The direction the first step the spiral takes when leaving the origin.
+        Should be one of up, down, left or right.
+    clockwise : bool, optional
+        If the spiral generated should be in a clockwise direction. Default
+        true.
+
+    Returns
+    -------
+    spiral_array : array, shape (dim, dim)
+        The spiral array.
+
+    Examples
+    --------
+    >>> generate_spiral(3, "up", True)
+    array([[8, 1, 2],
+           [7, 0, 3],
+           [6, 5, 4]], dtype=uint8)
+    """
+    if dim <= 0:
+        raise ValueError("dim must be a positive integer.")
+
+    if dim % 2 != 1:
+        raise ValueError("dim must be an odd integer.")
+
+    if direction not in ["down", "up", "left", "right"]:
+        raise ValueError("direction must be one of down, up, left or right.")
+
+    directions = {
+        "down": [1, 0],
+        "up": [-1, 0],
+        "left": [0, -1],
+        "right": [0, 1]
+    }
+
+    di, dj = directions[direction]
+    segment_length = 1
+    i = j = segment_passed = 0
```

Comment:
Can these statements be moved inside the for-loop and then just have to use `range(nrows * ncols)`?